File: sdks/dotnet/src/Croniq.Runner.Sdk/Protocol/WorkEvent.cs
using System.Text.Json.Serialization;

namespace Croniq.Runner.Sdk.Protocol;

/// <summary>
/// A structured log event pushed to a Croniq execution via
/// <c>POST /v1/work/{execution_id}/events</c>. Callers construct these
/// values (or use the helpers on <see cref="CroniqExecutionContext"/>); the SDK
/// auto-enriches the <see cref="Fields"/> map with <c>job_key</c>,
/// <c>runner_id</c>, and (when set) <c>runner_tags</c> before sending.
/// </summary>
public sealed record WorkEvent
{
    /// <summary>
    /// Severity level. Free-form string matching the server's accepted set
    /// (typically "trace", "debug", "info", "warn", "error"). When <c>null</c>,
    /// the server applies its default.
    /// </summary>
    [JsonPropertyName("level")]
    public string? Level { get; init; }

    /// <summary>Human-readable event text.</summary>
    [JsonPropertyName("message")]
    public required string Message { get; init; }

    /// <summary>
    /// Structured fields attached to the event. Existing keys win over
    /// SDK-injected ones — explicit caller values are preserved.
    /// </summary>
    [JsonPropertyName("fields")]
    public IReadOnlyDictionary<string, string>? Fields { get; init; }
}

File: sdks/java/core/src/main/java/io/croniq/runner/protocol/WorkEvent.java
package io.croniq.runner.protocol;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.Map;

/**
 * One log event streamed via POST /v1/work/{execution_id}/events.
 *
 * <p>The {@code fields} map carries caller-defined structured context.
 * Standard SDK-injected keys ({@code job_key}, {@code runner_id},
 * {@code runner_tags}) are added by the LogWriter in PR-4 — caller-supplied
 * values for those keys take precedence.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record WorkEvent(
        @JsonProperty("level") String level,
        @JsonProperty("message") String message,
        @JsonProperty("fields") Map<String, String> fields) {}
